File: 22.generate-parentheses.py
#
# @lc app=leetcode id=22 lang=python3
#
# [22] Generate Parentheses
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        def backtrack(S = '', left = 0, right = 0):
            if len(S) == 2 * n:
                ans.append(S)
                return
            if left < n:
                backtrack(S+'(', left+1, right)
            if right < left:
                backtrack(S+')', left, right+1)

        backtrack()
        return ans
'''回溯法 通过递归 实现 重点在于判断合理的组合的成立条件：
左括号数必须实时的或者说在任何时候都必须大于等于右括号数
否则就不合理 
# @lc code=end

File: 22.generate-parentheses.cpp
class Solution {
public:
    vector<string> ans={};
    void back(string s,int l,int r,int max)
    {
        
        if (s.size()==max*2) {
            ans.push_back(s);
            return ;
        }
        if(l<max) back(s+"(",l+1,r,max);
        
        if(r<l)  back(s+")",l,r+1,max);

    }
    vector<string> generateParenthesis(int n) {
  
        
        back("",0,0,n);
        return ans;

    }
};

/*官方java题解的c++做法 ，区别在于 在java题解中 使用的是引用传递而不是值传递 所以可以将
ans作为函数的一个参数加入递归函数中运行，但是在c++中 它是作为形参存在的，而形参的寿命就是
本函数以内，所以需要将ans作为一个全局变量声明在函数外，其他的好理解 。